src/vea/vea_alloc.c

```diff
@@ -395,50 +396,68 @@ persistent_alloc(struct vea_space_info *vsi, struct vea_free_extent *vfe)
 		return rc;
 	}
 
-	rc = verify_free_entry(&blk_off, &found);
+	found = (struct vea_free_extent *)val.iov_buf;
+	blk_off = (uint64_t *)key_out.iov_buf;
+
+	rc = verify_free_entry(blk_off, found);
 	if (rc)
 		return rc;
 
-	found_end = found.vfe_blk_off + found.vfe_blk_cnt;
+	found_end = found->vfe_blk_off + found->vfe_blk_cnt;
 	vfe_end = vfe->vfe_blk_off + vfe->vfe_blk_cnt;
 
-	if (found.vfe_blk_off > vfe->vfe_blk_off || found_end < vfe_end) {
+	if (found->vfe_blk_off > vfe->vfe_blk_off || found_end < vfe_end) {
 		D_ERROR("mismatched extent ["DF_U64", %u] ["DF_U64", %u]\n",
-			found.vfe_blk_off, found.vfe_blk_cnt,
+			found->vfe_blk_off, found->vfe_blk_cnt,
 			vfe->vfe_blk_off, vfe->vfe_blk_cnt);
 		return -DER_INVAL;
 	}
 
-	/* Remove the original free extent from persistent tree */
-	rc = dbtree_delete(btr_hdl, BTR_PROBE_EQ, &key_out, NULL);
-	if (rc)
-		return rc;
+	if (found->vfe_blk_off < vfe->vfe_blk_off) {
+		/* Adjust the in-tree free extent length */
+		rc = umem_tx_add_ptr(vsi->vsi_umem, &found->vfe_blk_cnt,
+				     sizeof(found->vfe_blk_cnt));
+		if (rc)
+			return rc;
+
+		found->vfe_blk_cnt = vfe->vfe_blk_off - found->vfe_blk_off;
 
-	/* Add back fore part of free extent */
-	if (found.vfe_blk_off < vfe->vfe_blk_off) {
-		frag = found;
-		frag.vfe_blk_cnt = vfe->vfe_blk_off - found.vfe_blk_off;
+		/* Add back the rear part of free extent */
+		if (found_end > vfe_end) {
+			frag.vfe_blk_off = vfe->vfe_blk_off + vfe->vfe_blk_cnt;
+			frag.vfe_blk_cnt = found_end - vfe_end;
+			rc = daos_gettime_coarse(&frag.vfe_age);
+			if (rc)
+				return rc;
 
-		d_iov_set(&key_in, &frag.vfe_blk_off,
-			     sizeof(frag.vfe_blk_off));
-		d_iov_set(&val, &frag, sizeof(frag));
-		rc = dbtree_update(btr_hdl, &key_in, &val);
+			d_iov_set(&key_in, &frag.vfe_blk_off,
+				  sizeof(frag.vfe_blk_off));
+			d_iov_set(&val, &frag, sizeof(frag));
+			rc = dbtree_update(btr_hdl, &key_in, &val);
+			if (rc)
+				return rc;
+		}
+	} else if (found_end > vfe_end) {
+		/* Adjust the in-tree integer key */
```

Comment:
so this is safe because it will not change the ordering in the tree?

Reply:
Right, I think so.